Battleonly mode

Laserlicht opened the pull request:
Currently supported:
- Battle against heroes
- Battle against hero in town
- Battle against town without hero
- Selecting hero
- Selecting creature
- Selecting primary skill level
- Selecting battleground

Selectable in single player at map selection (small icon on top left).

Progress is tracked here: #5303

IvanSavenko commented on Mods/vcmi/Content/config/english.json:
```diff
@@ -138,6 +138,12 @@
 	"vcmi.lobby.deleteMode" : "Switch to delete mode and back",
+	"vcmi.lobby.battleOnlyMode" : "Battle only mode",
+	"vcmi.lobby.battleOnlyModeBattleground" : "Battleground",
+	"vcmi.lobby.battleOnlyModeBattlegroundSelect" : "Select Battleground",
```
Should be called "battlefield", not "battleground"

Can be replaced everywhere - e.g. via search-and-replace

IvanSavenko commented on client/CPlayerInterface.cpp:
```diff
@@ -658,7 +659,7 @@ void CPlayerInterface::battleStart(const BattleID & battleID, const CCreatureSet
 	EVENT_HANDLER_CALLED_BY_CLIENT;
 
-	bool useQuickCombat = settings["adventure"]["quickCombat"].Bool();
+	bool useQuickCombat = settings["adventure"]["quickCombat"].Bool() || GAME->map().getMap()->battleOnly;
```
Shouldn't it be the opposite? 

If game is in battle-only mode, never use quick combat and always manual combat?

Laserlicht replied:
I prefer it this way because you can check if you are better (with fewer losses) than AI.

Laserlicht replied:
Yes. For first time. But I think it's easy to understand the second time.

Fheroes2 uses the same approach.

Laserlicht replied:
May add a checkbox to disable/enable this in a later PR.

IvanSavenko commented on lib/modding/IdentifierStorage.cpp:
```diff
@@ -126,7 +126,7 @@ CIdentifierStorage::ObjectCallback CIdentifierStorage::ObjectCallback::fromNameW
 CIdentifierStorage::ObjectCallback CIdentifierStorage::ObjectCallback::fromNameAndType(const std::string & scope, const std::string & type, const std::string & fullName, const std::function<void(si32)> & callback, bool optional, bool bypassDependenciesCheck, bool caseSensitive)
 {
-	assert(!scope.empty());
+	//assert(!scope.empty());
```
Leftover from debug? This generally means corrupted request

Same for another edit in this file

IvanSavenko commented on client/lobby/SelectionTab.cpp:
```diff
@@ -325,6 +335,7 @@ void SelectionTab::toggleMode()
 				inputName->disable();
 				auto files = getFiles("Maps/", EResType::MAP);
 				files.erase(ResourcePath("Maps/Tutorial.tut", EResType::MAP));
+				files.erase(ResourcePath("Maps/BattleOnlyMode.vmap", EResType::MAP));
```
What is this map?

IvanSavenko commented on client/lobby/BattleOnlyMode.cpp:
```diff
@@ -0,0 +1,461 @@
+		}
+
+		creatureImage[i]->addLClickCallback([this, i](){
+			auto creatures = LIBRARY->creh->objects;
```
Same here. Better to use LIBRARY->creh->getDefaultAllowed call and iterate over creature ID's

IvanSavenko commented:
Also - can you create issue with your plans for this feature (e.g. MP support) to have proper place to discuss final form / suggestions for this feature that are out of scope for this PR?

Laserlicht commented on lib/bonuses/CBonusSystemNode.cpp:
```diff
@@ -184,7 +184,6 @@ CBonusSystemNode::~CBonusSystemNode()
 
 void CBonusSystemNode::attachTo(CBonusSystemNode & parent)
 {
-	assert(!vstd::contains(parentsToPropagate, &parent));
```
It seems the bonus system does not work with `EditorCallback` (maybe not implemented gamestate?).
It's needed for fighting on boat on water. without removing the asserts are triggered. No idea why.

But after removing - it's fully working.

IvanSavenko replied:
When exactly these asserts are triggered? Can you show call stack? Looks like double-attach.

Previously, we had weird bugs where bonus system ended up in invalid state, like entities that were attached to wrong parents. Removing these asserts means that such bugs can resurface again

Laserlicht commented on client/lobby/BattleOnlyMode.cpp:
```diff
@@ -0,0 +1,461 @@
+				index--;
+				auto creature = creatures[index];
+				int amount = selectedArmy->slotEmpty(SlotID(i)) ? 1 : selectedArmy->Slots().at(SlotID(i))->getCount();
+				ENGINE->windows().createAndPushWindow<NumberInputWindow>(amount, [this, creature, i](int quantity){
```
Yes. Also considered this. But identified it as improvement area for later. Because needs to adjust assets for fitting boxes.

Laserlicht commented:
> For what is used right side? I have tried to set battle and it worked - must click on manual battle as it uses quick combat right now as default.

Right area is for setting artifacts, secondary skills, spell selection, additional buttons (like "reset") and maybe other ideas.
Not sure when it comes in action. And if it is maybe better to drop right area and instead add custom buttons (more flexible).

> Also, will be possible to use something like "start battleonly" in [mainmenu.json](https://github.com/vcmi/vcmi/blob/6b4179af86fae2bb5a78db20a539aba05d42cb95/config/mainmenu.json#L78)? Seems like we can add it to custom main menu pack when we have proper button image - probably two titans swords crossed with "Quick battle" text?

Also considered it. But I think also for later (not really on the top of priority).

Laserlicht commented:
> Also - can you create issue with your plans for this feature (e.g. MP support) to have proper place to discuss final form / suggestions for this feature that are out of scope for this PR?

Yes we can use #5303. Can add a todo list later.

Laserlicht commented:
Added todo here: #5303

Changed layout and added reset button.
<img width="851" height="653" alt="grafik" src="https://github.com/user-attachments/assets/12ef6823-039b-4fec-912b-bb54434f029b" />

Also improved numeric input.

Open review points will be fixed later.

IvanSavenko commented on lib/texts/TextOperations.h:
```diff
@@ -116,4 +121,67 @@ inline std::string TextOperations::formatMetric(Arithmetic number, int maxDigits
+template<typename Arithmetic>
+inline Arithmetic TextOperations::parseMetric(const std::string &text)
+{
+	static const std::string symbols = " kMGTPE";
```
Unused?
```suggestion
```

IvanSavenko commented:
Something is off:
- Title bar is named simply "creature", should have been "Select Creatures"
- What's up with sorting? It is neither alphabetical, nor sorted by index. Ideally should be faction / level / upgrade
<img width="633" height="948" alt="зображення" src="https://github.com/user-attachments/assets/b67eeed7-9e63-41ff-8d41-14930fe97223" />

IvanSavenko commented:
Battle only (or something similar, like battle only mode) is fine IMO.

Personally would've called it "Duel mode", but in H3 "Duel" often refers to that RMG template
